bin/setup

```diff
@@ -1,41 +1,106 @@
+    run "bin/rake log:clear tmp:clear"
+    puts "\n== Restarting application server =="
+    run  "bin/rails restart"
+  end
+end
+
+require "fileutils"
+begin
+  require "highline"
+rescue LoadError
+  # If highline is missing, we'll gracefully omit ansi color output
+end
+
 include FileUtils
 
 # path to your application root.
-APP_ROOT = File.expand_path('..', __dir__)
+def within_project_root(&block)
+  Dir.chdir(File.expand_path("../..", __FILE__), &block)
+end
 
-def system!(*args)
-  system(*args) || abort("\n== Command #{args} failed ==")
+def test(opts)
+  command, output = opts.first
+
+  log(:yellow, "test #{command}") do
+    unless `#{command}`.include?(output)
+      die("#{command} does not include #{output}")
+    end
+  end
 end
 
-chdir APP_ROOT do
-  # This script is a starting point to setup your application.
-  # Add necessary setup steps to this file.
+def ruby_version
+  @ruby_version ||= begin
+    IO.read(".ruby-version").strip
+  end
+end
 
-  puts '== Installing dependencies =='
-  system! 'gem install bundler --conservative'
-  system('bundle check') || system!('bundle install')
+def run(command)
+  log(:blue, "run  #{command}") do
+    shell = "#{command} > /dev/null"
+    with_clean_bundler_env do
+      system(shell) or die("#{command} exited with non-zero status")
+    end
+  end
+end
 
-  # Install JavaScript dependencies if using Yarn
-  system('bin/yarn')
+def with_clean_bundler_env(&block)
+  return block.call unless defined?(Bundler)
+  Bundler.with_clean_env(&block)
+end
+
+def copy(example_path)
+  copy_path = example_path.sub(/\.?example/, "")
+  return if File.exist?(copy_path)
 
-  unless File.exist?('.env')
-    puts "\n== Creating development .env file =="
-    cp 'env-example', '.env'
-    print "\n Generating SECRET_KEY_BASE..."
-    system "echo 'SECRET_KEY_BASE: '`rake secret`'\\n' >> .env"
-    puts "√"
+  log(:magenta, "copy #{example_path} → #{copy_path}") do
+    FileUtils.cp(example_path, copy_path)
   end
+end
 
-  puts "\n== Preparing database =="
-  system! 'bin/rails db:setup'
+def test_local_env_contains_required_keys
+  keys = ->(f) { IO.readlines(f).map { |l| l[/^([^#\s][^=\s]*)/, 1] }.compact }
 
-  puts "\n== Removing old logs and tempfiles =="
-  system! 'bin/rails log:clear tmp:clear'
+  log(:yellow, "test .env contents") do
+    missing = keys["example.env"] - keys[".env"]
+    if missing.any?
+      die("Your .env file is missing #{missing.join(', ')}")
+    end
+  end
+end
 
-  puts "\n== Restarting application server =="
-  system! 'bin/rails restart'
+def log(color, message, out=$stdout, &block)
+  if defined?(HighLine::String)
+    message.sub!(/^(\S*)/) { HighLine::String.new($1).public_send(color) }
+  end
+
+  if block_given?
+    out.print("#{message}… ")
+    yield
+    log(:green, "✔︎")
+  else
+    out.puts(message)
+  end
 end
+
+def die(message)
+  puts
+  log(:red, "FAIL #{message}", $stderr)
+  exit(1)
+end
+
+setup!
```

Comment:
This is a lot of changes - is it copied from somewhere, or have you written it all new?

Reply:
From Croissant + some of the previous code. :stuck_out_tongue: